test/cpp/jit/test_gpu.cpp

```diff
@@ -15168,6 +15190,1569 @@ TEST(NVFuserTest, FusionZeroSizeTensorNormalization_CUDA) {
+        int rhs_value = *rhs->value();
+        TORCH_CHECK(lhs_value == split_factor && rhs_value == 2);
+      }
+    }
+  }
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  int numel_x = 99;
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({numel_x}, options);
+  std::vector<IValue> inputs = {t0};
+  auto outputs = fe.runFusion(inputs);
+
+  auto ref = (t0 + shift(t0, {-1}) + shift(t0, {1})) / 3;
+
+  testValidate(&fusion, outputs, inputs, {ref}, __LINE__, __FILE__);
+}
+
+TEST(NVFuserTest, FusionShift5ptStencil_CUDA) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  // 5-pt stencil
+  auto tv0 = makeSymbolicTensor(2);
+  fusion.addInput(tv0);
+  std::vector<std::vector<int>> offsets = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+  std::vector<TensorView*> tvs;
+  for (const auto& offset : offsets) {
+    tvs.push_back(shift(tv0, offset));
+  }
+
+  auto tv_out = tv0;
+
+  for (auto tv : tvs) {
+    tv_out = add(tv_out, tv);
+  }
+
+  tv_out = div(tv_out, new Double(tvs.size() + 1));
+
+  fusion.addOutput(tv_out);
+
+  std::vector<int> split_factor({4, 8});
+
+  tv_out->split(-1, split_factor[1]);
+  tv_out->split(0, split_factor[0]);
+  tv_out->reorder({{1, 2}, {2, 1}});
+
+  auto cache = tv0->cache_after();
+
+  tv0->computeAt(tv_out, 2);
+
+  // Inline completely except for the cache
+  for (auto tv : tvs) {
+    tv->computeAt(tv_out, -1);
+  }
+
+  // cache allocation: (split_factor + 2) * (split_factor + 2)
+  GpuLower gpulw(&fusion);
+  for (const auto& kir_node : gpulw.kernel()->irNodes()) {
+    if (auto alloc = dynamic_cast<kir::Allocate*>(kir_node.get())) {
+      auto tensor_name = alloc->buffer()->name();
+      if (tensor_name == cache->name()) {
+        TORCH_CHECK(alloc->shape().size() == 2);
+        for (int i = 0; i < 2; ++i) {
+          auto def =
+              dynamic_cast<kir::BinaryOp*>(alloc->shape().at(i)->definition());
+          auto lhs = dynamic_cast<kir::Int*>(def->as<kir::BinaryOp>()->lhs());
+          TORCH_CHECK(lhs != nullptr && lhs->isConst());
+          int lhs_value = *lhs->value();
+          auto rhs = dynamic_cast<kir::Int*>(def->as<kir::BinaryOp>()->rhs());
+          TORCH_CHECK(rhs != nullptr && rhs->isConst());
+          int rhs_value = *rhs->value();
+          TORCH_CHECK(lhs_value == split_factor[i] && rhs_value == 2);
+        }
+      }
+    }
+  }
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  int numel_x = 99;
+  int numel_y = 101;
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({numel_x, numel_y}, options);
+  std::vector<IValue> inputs = {t0};
+  auto outputs = fe.runFusion(inputs);
+
+  auto ref = t0;
+  for (const auto& offset : offsets) {
+    ref = ref + shift(t0, offset);
+  }
+  ref = ref / int(offsets.size() + 1);
+
+  testValidate(&fusion, outputs, inputs, {ref}, __LINE__, __FILE__);
+}
+
+TEST(NVFuserTest, FusionShift9ptStencil_CUDA) {
```

Comment:
I think that would be more suitable for actual users. Shift is versatile but may be considered too primitive in practice.

Reply:
Sure, I also thought maybe if we had something like a 7x7 convolution it would be burdensome to use, or even if not it could blow out the instruction cache.